File: exercise/python/day22/eid/serial_number.py
from typing import Union

from eid.parsing import parse_or_error
from eid.parsing_error import ParsingError

PARSING_ERROR_MESSAGE = "Serial number should be gt 0 and lt 1000"


def is_valid(value):
    return 1 <= value <= 999


class SerialNumber:
    def __init__(self, value: int):
        if not is_valid(value):
            raise ValueError(PARSING_ERROR_MESSAGE)
        self.value = value

    def __str__(self):
        return f"{self.value:03}"


def parse_serial_number(text: str) -> Union[ParsingError, SerialNumber]:
    return parse_or_error(text, lambda x: is_valid(x), SerialNumber, PARSING_ERROR_MESSAGE)

File: exercise/java/day22/src/main/java/eid/SerialNumber.java
package eid;

import io.vavr.control.Either;
import lombok.EqualsAndHashCode;
import lombok.experimental.ExtensionMethod;

@ExtensionMethod(StringExtensions.class)
@EqualsAndHashCode
public class SerialNumber {
    private final int value;

    private SerialNumber(int value) {
        this.value = value;
    }

    public static SerialNumber fromInt(int potentialSerialNumber) {
        return new SerialNumber(potentialSerialNumber);
    }

    public static Either<ParsingError, SerialNumber> parseSerialNumber(String potentialSerialNumber) {
        return potentialSerialNumber
                .toInt()
                .filter(x -> x >= 1 && x <= 999)
                .map(SerialNumber::new)
                .toEither(new ParsingError("serial number should be between 1 and 999"));
    }

    @Override
    public String toString() {
        return String.format("%03d", value);
    }
}